networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRangeMessageHandler.java

```diff
@@ -105,24 +105,47 @@ public void onIncomingMessage(
     final UInt64 count = maxRequestSize.min(message.getCount());
     final UInt64 endSlot = message.getStartSlot().plus(message.getStep().times(count)).minus(ONE);
 
-    final UInt64 headBlockSlot =
-        combinedChainDataClient.getBestBlock().map(SignedBeaconBlock::getSlot).orElse(ZERO);
-    final NavigableMap<UInt64, Bytes32> hotRoots;
-    if (combinedChainDataClient.isFinalized(endSlot)) {
-      // All blocks are finalized so skip scanning the protoarray
-      hotRoots = new TreeMap<>();
-    } else {
-      hotRoots =
-          combinedChainDataClient.getAncestorRoots(
-              message.getStartSlot(), message.getStep(), count);
-    }
-    // Don't send anything past the last slot found in protoarray to ensure blocks are consistent
-    // If we didn't find any blocks in protoarray, every block in the range must be finalized
-    // so we don't need to worry about inconsistent blocks
-    final UInt64 headSlot = hotRoots.isEmpty() ? headBlockSlot : hotRoots.lastKey();
-    return sendNextBlock(
-        new RequestState(
-            message.getStartSlot(), message.getStep(), count, headSlot, hotRoots, callback));
+    return combinedChainDataClient
+        .getEarliestHistoricalBlockSlot()
+        .thenCompose(
+            earliestSlot -> {
+              if (earliestSlot.map(s -> s.isGreaterThan(message.getStartSlot())).orElse(true)) {
+                // We're missing the first block so return an error
+                return SafeFuture.failedFuture(
+                    new RpcException.HistoricalDataUnavailableException(
+                        "Requested historical blocks are currently unavailable"));
```

Comment:
Can you please comment why do we need to return the error instead of empty set? Doesn't this error mean that we totally have no any historical blocks (aka light client)?

Reply:
Initially I had this returning an empty set.  But the problem is similar to the one mentioned above:  the receiving node could interpret this to mean that all of the blocks in this range were skipped.  It will be easier to write sync logic if we can disambiguate missing data from skipped blocks.

Reply:
We'll hit this case if we start up from a recent state.  Initially, we won't have any blocks prior to that state.  Planning to implement a backwards sync that will pull historical blocks prior to this state, but it will still take time to get all of those blocks so we need to handle the case where they're missing.

Reply:
Generally I think that a requesting client shouldn't interpret the absence of blocks in a response as 'empty slots'. I think that one may consider the slots are indeed empty when it has two chained blocks (child -> parent) bounding those slots. (i.e. if and only if one see `block[100] <- block[90]` he may consider slots `91...99` as empty)

I didn't find any strict rules in the spec with respect to this situation, but my hypothesis is that we should response with empty list. The similar (though not exact the same) way we would response with empty list if we are still syncing and have no requested blocks yet. IMO generally responding with error should denote some exceptional situation. 

Anyways this case seems worth clarifying in the spec

Reply:
BTW the exact chosen code `128` makes me think that probably other client uses it the same way?

Reply:
Yeah, we should clarify this in the spec - planning to open an issue or a PR.

The practical difference wrt the initial sync and missing recent blocks is that we have the `Status` message from our peers that tells us how far the chain has synced.  But there is no information on how much historical chain data is available.  I think ideally we would start sending the range of available blocks in the `Status` message, but we don't have that at the moment.

Reply:
128 is the start of the custom error code range:

> Clients MAY use response codes above 128 to indicate alternative, erroneous request-specific responses.
>
> The range [3, 127] is RESERVED for future usages, and should be treated as error if not recognized expressly.

Reply:
Just one more comment: I suppose that a client is free to respond with _any_ subset of valid blocks which satisfy request criteria

Reply:
The spec has the following constraints:
>Clients MUST respond with at least the first block that exists in the range, if they have it, and no more than MAX_REQUEST_BLOCKS blocks.
>
>The following blocks, where they exist, MUST be sent in consecutive order.

Which I interpret to mean that you need to start at the first block in the range, but that isn't 100% clear.

Reply:
> Which I interpret to mean that you need to start at the first block in the range, but that isn't 100% clear.

Yep looks like this statement may allow different treatments 
`..., if they have it, ...` :)